src/apps/lwaftr/fragmentv4.lua

```diff
@@ -0,0 +1,289 @@
+module(..., package.seeall)
+
+local constants = require("apps.lwaftr.constants")
+local lwutil = require("apps.lwaftr.lwutil")
+local packet = require("core.packet")
+local bit = require("bit")
+local ffi = require("ffi")
+
+local rd16, wr16, get_ihl = lwutil.rd16, lwutil.wr16, lwutil.get_ihl
+local cast = ffi.cast
+local C = ffi.C
+local band, bor = bit.band, bit.bor
+local ceil = math.ceil
+
+local total_length_offset = constants.ethernet_header_size + constants.o_ipv4_total_length
+local frag_id_offset = constants.ethernet_header_size + constants.o_ipv4_identification
+local flags_and_frag_offset_offset = constants.ethernet_header_size + constants.o_ipv4_flags
+
+-- Constants to manipulate the flags next to the frag-offset field directly
+-- as a 16-bit integer, without needing to shift the 3 flag bits.
+local flag_dont_fragment_mask  = 0x4000
+local flag_more_fragments_mask = 0x2000
+local frag_offset_field_mask   = 0x1FFF
+
+
+-- TODO: Consider security/performance tradeoffs of randomization
+local fresh_frag_id = (function ()
+   local internal_frag_id = 0x4242
+   return function ()
+      internal_frag_id = band(internal_frag_id + 1, 0xFFFF)
+      return internal_frag_id
+   end
+end)()
+
+
+FRAGMENT_OK = 1
+FRAGMENT_UNNEEDED = 2
+FRAGMENT_FORBIDDEN = 3
+
+--
+-- IPv4 fragmentation, as per https://tools.ietf.org/html/rfc791
+--
+-- For an invocation:
+--
+--    local statuscode, packets = fragment_ipv4(input_packet, mtu)
+--
+-- the possible values for the returned "statuscode" are:
+--
+--   * FRAGMENT_OK: the returned "packets" is a list of IPv4 packets, all
+--     of them smaller or equal than "mtu" bytes, which contain the payload
+--     from the "input_packet" properly fragmented. Note that "input_packet"
+--     is modified in-place.
+--
+--   * FRAGMENT_UNNEEDED: the returned "packets" is the same object as
+--     "input_packet", unmodified. This is the case when the size of packet
+--     is smaller or equal than "mtu" bytes.
+--
+--   * FRAGMENT_FORBIDDEN: the returned "packets" will be "nil". This is
+--     the case when "input_packet" has the "don't fragment" flag set, and
+--     its size is bigger than "mtu" bytes. Client code may want to return
+--     an ICMP Datagram Too Big (Type 3, Code 4) packet back to the sender.
+--
+function fragment_ipv4(ipv4_pkt, mtu)
+   if ipv4_pkt.length <= mtu then
+      return FRAGMENT_UNNEEDED, ipv4_pkt
+   end
+
+   -- Discard packets with the DF (dont't fragment) flag set
+   do
+		local flags_and_frag_offset = C.ntohs(rd16(ipv4_pkt.data + flags_and_frag_offset_offset))
```

Comment:
I'm afraid some tabs slipped in. SnS uses 3-space indents for Lua code, no tabs. I have the git pre-commit hook activated, so it prevents me from commiting tabs, trailing white spaces, etc in the source code.